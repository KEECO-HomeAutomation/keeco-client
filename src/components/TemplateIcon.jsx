import React from 'react';
import PropTypes from 'prop-types';
import { compose, mapProps } from 'recompose';

import DeviceHubIcon from '@material-ui/icons/DeviceHub';

export const TemplateIcon = ({ Icon, ...rest }) => {
	return <Icon {...rest} />;
};

TemplateIcon.propTypes = {
	Icon: PropTypes.elementType
};

export const enhancer = compose(
	mapProps(({ template, templateData, ...rest }) => {
		let icon;
		switch (template) {
			case 'switch':
				if (templateData.on) {
					icon = DeviceHubIcon;
				} else {
					icon = DeviceHubIcon;
				}
				break;
			case 'lamp':
				icon = DeviceHubIcon;
				break;
			case 'thermostat':
				icon = DeviceHubIcon;
				break;
			default:
				icon = DeviceHubIcon;
				break;
		}
		return {
			Icon: icon,
			...rest
		};
	})
);

export default enhancer(TemplateIcon);
