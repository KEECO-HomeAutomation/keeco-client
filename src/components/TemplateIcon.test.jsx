/* eslint-disable react/display-name */
/* eslint-disable react/prop-types */
import React from 'react';
import { mount } from 'enzyme';

jest.mock('@material-ui/icons/DeviceHub', () => () => <p>DeviceHubIcon</p>);

import TemplateIcon, { TemplateIcon as Base, enhancer } from './TemplateIcon';

import DeviceHubIcon from '@material-ui/icons/DeviceHub';

describe('<TemplateIcon />', () => {
	test('Should export base and enhancer', () => {
		expect(Base).not.toBe(undefined);
		expect(enhancer).not.toBe(undefined);
	});

	test('Should render without errors', () => {
		const comp = mount(<TemplateIcon />);
		expect(comp.length).toBe(1);
	});

	test('Should match snapshot', () => {
		expect(mount(<TemplateIcon />)).toMatchSnapshot();
	});

	describe('Should render the matching icon', () => {
		const testCases = [
			{ name: 'switch', Comp: DeviceHubIcon },
			{ name: 'lamp', Comp: DeviceHubIcon },
			{ name: 'thermostat', Comp: DeviceHubIcon }
		];
		testCases.forEach(testCase => {
			test(
				'When template=' + testCase.name + ' should render the matching icon',
				() => {
					const comp = mount(
						<TemplateIcon template={testCase.name} templateData={{}} />
					);
					expect(comp).toMatchSnapshot();
					expect(comp).toContainReact(<testCase.Comp />);
				}
			);
		});
	});

	test('When passed a non handled template should render the standard icon', () => {
		const comp = mount(
			<TemplateIcon template={'notExistingTemplateName'} templateData={{}} />
		);
		expect(comp).toMatchSnapshot();
		expect(comp).toContainReact(<DeviceHubIcon />);
	});

	describe('Handle templateData', () => {
		describe('template=switch', () => {
			test('on=true', () => {
				const comp = mount(
					<TemplateIcon template="switch" templateData={{ on: true }} />
				);
				expect(comp).toMatchSnapshot();
			});
			test('on=false', () => {
				const comp = mount(
					<TemplateIcon template="switch" templateData={{ on: false }} />
				);
				expect(comp).toMatchSnapshot();
			});
		});
	});
});
