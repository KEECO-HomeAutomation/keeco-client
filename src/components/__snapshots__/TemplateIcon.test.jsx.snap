// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TemplateIcon /> Handle templateData template=switch on=false 1`] = `
<mapProps(TemplateIcon)
  template="switch"
  templateData={
    Object {
      "on": false,
    }
  }
>
  <TemplateIcon
    Icon={[Function]}
  >
    <Component>
      <p>
        DeviceHubIcon
      </p>
    </Component>
  </TemplateIcon>
</mapProps(TemplateIcon)>
`;

exports[`<TemplateIcon /> Handle templateData template=switch on=true 1`] = `
<mapProps(TemplateIcon)
  template="switch"
  templateData={
    Object {
      "on": true,
    }
  }
>
  <TemplateIcon
    Icon={[Function]}
  >
    <Component>
      <p>
        DeviceHubIcon
      </p>
    </Component>
  </TemplateIcon>
</mapProps(TemplateIcon)>
`;

exports[`<TemplateIcon /> Should match snapshot 1`] = `
<mapProps(TemplateIcon)>
  <TemplateIcon
    Icon={[Function]}
  >
    <Component>
      <p>
        DeviceHubIcon
      </p>
    </Component>
  </TemplateIcon>
</mapProps(TemplateIcon)>
`;

exports[`<TemplateIcon /> Should render the matching icon When template=lamp should render the matching icon 1`] = `
<mapProps(TemplateIcon)
  template="lamp"
  templateData={Object {}}
>
  <TemplateIcon
    Icon={[Function]}
  >
    <Component>
      <p>
        DeviceHubIcon
      </p>
    </Component>
  </TemplateIcon>
</mapProps(TemplateIcon)>
`;

exports[`<TemplateIcon /> Should render the matching icon When template=switch should render the matching icon 1`] = `
<mapProps(TemplateIcon)
  template="switch"
  templateData={Object {}}
>
  <TemplateIcon
    Icon={[Function]}
  >
    <Component>
      <p>
        DeviceHubIcon
      </p>
    </Component>
  </TemplateIcon>
</mapProps(TemplateIcon)>
`;

exports[`<TemplateIcon /> Should render the matching icon When template=thermostat should render the matching icon 1`] = `
<mapProps(TemplateIcon)
  template="thermostat"
  templateData={Object {}}
>
  <TemplateIcon
    Icon={[Function]}
  >
    <Component>
      <p>
        DeviceHubIcon
      </p>
    </Component>
  </TemplateIcon>
</mapProps(TemplateIcon)>
`;

exports[`<TemplateIcon /> When passed a non handled template should render the standard icon 1`] = `
<mapProps(TemplateIcon)
  template="notExistingTemplateName"
  templateData={Object {}}
>
  <TemplateIcon
    Icon={[Function]}
  >
    <Component>
      <p>
        DeviceHubIcon
      </p>
    </Component>
  </TemplateIcon>
</mapProps(TemplateIcon)>
`;
